third-party/rsutils/src/shorten-json-string.cpp

```diff
@@ -0,0 +1,128 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2023 Intel Corporation. All Rights Reserved.
+
+#include <rsutils/string/shorten-json-string.h>
+
+
+namespace rsutils {
+namespace string {
+
+
+// Given an outside block, e.g.:
+//        012345678901234567890123456789012345678901234567890123456789012345678901234567890
+//        0         1         2         3         4         5         6         7         8
+//        {"one":1,"two":[1,2,3],"three":{"longassblock":{"insideblock":89012}},"four":4}
+// Find the first inside block, including the enclosing delimiters (parenthesis or square brackets):
+//                       ^______^        ^_____________________________________^
+//
+static slice find_inside_block( slice const & outside )
+{
+    // find an opening
+    slice::const_iterator it = outside.begin() + 1;  // opening {, separating comma, etc.
+    bool in_quote = false;
+    while( it < outside.end() && ( in_quote || *it != '[' && *it != '{' ) )
+    {
+        if( *it == '"' )
+            in_quote = ! in_quote;
+        ++it;
+    }
+    if( it >= outside.end() )
+        return slice();
+    assert( ! in_quote );
+    auto const begin = it;
+    char const open = *begin;
+
+    // find the closing
+    char const close = ( *it == '[' ) ? ']' : '}';
+    int nesting = 0;
+    while( 1 )
+    {
+        if( ++it == outside.end() )
+            // Something is invalid
+            return slice();
+
+        if( *it == '"' )
+            in_quote = ! in_quote;
+        else if( ! in_quote )
+        {
+            if( *it == close )
+            {
+                if( 0 == nesting )
+                    break;
+                --nesting;
+            }
+            else if( *it == open )
+                ++nesting;
+        }
+    }
+    return slice( begin, it + 1 );  // including the enclosing {} or []
+}
+
+
+// max-length output
+// ---------- ---------------------------------------------------
+//          7 { ... }
+//          8 {" ... }
+//          9 {"a ... }
+//         22 {"a[1]":1,"b[2": ... }
+//         29 {"a[1]":1,"b[2":3,"d":[ ... }
+//         41 {"a[1]":1,"b[2":3,"d":[1,2,{3,4,5, ... ]}
+//         42 {"a[1]":1,"b[2":3,"d":[1,2,{ ... },6,7,8]}
+//         43 {"a[1]":1,"b[2":3,"d":[1,2,{3 ... },6,7,8]}
+//         49 {"a[1]":1,"b[2":3,"d":[1,2,{3,4,5,6 ... },6,7,8]}
+//         50 {"a[1]":1,"b[2":3,"d":[1,2,{3,4,5,6,7,8,9},6,7,8]}   <--   original json string
+//
+ellipsis shorten_json_string( slice const & str, size_t max_length )
+{
+    if( str.length() <= max_length )
+        // Already short enough
+        return ellipsis( str, slice() );
+    if( max_length < 7 )
+        return ellipsis( slice(), str );  // impossible: "{ ... }" is the minimum
+
+    // Try to find an inside block that can be taken out
+    ellipsis final;
```

Comment:
`final` can serve as a C++ keyword in other context, better to use other name. `result`?

Reply:
fixed